### PR TITLE
Migrate to a hybrid of GH actions and Travis

### PR DESCRIPTION
**Pull request by GrahamCampbell**
Removes circle and appveyor, and some of the travis jobs. Long term, we could probably remove travis, but this is not necessary yet. The purpose of this PR is to have fast robust testing across multiple PHP versions and OSs, and not for overkill consolidation, taking extra developer time without significant benefit.

### CODE REVIEW

File: phpunit.xml.dist [hunk 1]
```diff
@@ -53,6 +53,7 @@
     <php>
         <ini name="zend.enable_gc" value="0"/>
         <ini name="memory_limit" value="1G"/>
+        <env name="FAST_LINT_TEST_CASES" value="0"/>
```
Comment (julienfalque) [inline phpunit.xml.dist:56]: What is the purpose of this variable? I mean: if it only changes the time required to lint files, shouldn't we always enable it? Or is there any trade-off when doing so?

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+      - name: Setup problem matchers
+        run: echo "::add-matcher::${{ runner.tool_cache }}/phpunit.json"
+
+      - name: Install platform ignoring dependencies
+        uses: nick-invision/retry@v1
+        with:
+          timeout_minutes: 5
+          max_attempts: 5
+          command: composer update --prefer-dist --no-interaction --no-progress --ignore-platform-req=php
```
Comment (julienfalque) [inline .github/workflows/tests.yml:46]: I would remove `--no-progress` so we can see what happens if needed.

Is `--no-interaction` required at all?

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+          max_attempts: 5
+          command: composer update --prefer-stable --prefer-lowest --prefer-dist --no-interaction --no-progress
+        if: "matrix.stability == 'Lowest'"
+
+      - name: Execute tests
+        run: vendor/bin/phpunit
+        if: "matrix.php != '8.0'"
+
+      - name: Execute tests with weak deprecation warnings
+        run: SYMFONY_DEPRECATIONS_HELPER=weak vendor/bin/phpunit
```
Comment (julienfalque) [inline .github/workflows/tests.yml:70]: In my tests I didn't need this `SYMFONY_DEPRECATIONS_HELPER=weak` variable to make PHPUnit pass, are you sure it is required?

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        if: "matrix.php == '8.0'"
+
+      - name: Execute php-cs-fixer
+        run: PHP_CS_FIXER_FUTURE_MODE=1 PHP_CS_FIXER_IGNORE_ENV=1 php php-cs-fixer --diff --dry-run -v fix
```
Comment (julienfalque) [inline .github/workflows/tests.yml:74]: Can we have `PHP_CS_FIXER_IGNORE_ENV=1` for PHP 8.0 jobs only?

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+  windows_tests:
```
Comment (julienfalque) [inline .github/workflows/tests.yml:76]: Can we make the OS part of the matrix of the first job instead of having multiple job definitions? It seems it would prevent at lot of duplicated steps.

**Review by julienfalque — COMMENTED** [review]
Damn, I was working on this too, but you were faster than me :)

File: tests/Fixtures/Integration/set/@PHP8Migration.test [hunk 6]
```diff
@@ -1,6 +0,0 @@
---TEST--
```
[the remark this answers is not in the document]
Comment (julienfalque) [reply tests/Fixtures/Integration/set/@PHP8Migration.test:1]: Actually this is not from a commit of yours, maybe you need to rebase?

File: phpunit.xml.dist (hunk 1 above)
Comment (GrahamCampbell) [reply phpunit.xml.dist:56]: All these changes are from another PR. Please review that one. I will rebase this after that has merged. ;)

NB The reason for fast lint was because the tests took over an hour otherwise after the full linting was added. The reason for the full linting is explained in my other PR.

File: tests/Fixtures/Integration/set/@PHP8Migration.test (hunk 6 above)
[the remark this answers is not in the document]
Comment (GrahamCampbell) [reply tests/Fixtures/Integration/set/@PHP8Migration.test:1]: Yes, this is because this PR is WIP and contains commits from my other PR joint with @SpacePossum. These will disappear after that is merged and this PR is rebased. ;)

File: .github/workflows/tests.yml (hunk 2 above)
Comment (GrahamCampbell) [reply .github/workflows/tests.yml:46]: No progress doesn't "hide progress" it hides the dynamic "process bars". The steps are still shown.

File: .github/workflows/tests.yml (hunk 3 above)
Comment (GrahamCampbell) [reply .github/workflows/tests.yml:70]: This is needed for PHP 8.0. The UI looks good with just warnings, but if you check the error code, you will find it is non-zero. ;)

File: .github/workflows/tests.yml (hunk 5 above)
Comment (GrahamCampbell) [reply .github/workflows/tests.yml:76]: Each OS has different steps, and the duplication is least horrible this way, I feel. We could make the file shorter, but this would make it less readable for sure. Ultimately, we want everyone to be able to easily contribute to this file.

Comment (julienfalque) [reply .github/workflows/tests.yml:76]: Agreed, the Travis CI config file is currectly very hard to reason about. Yet, I feel we can still have a single matrix hurting readability too much. Maybe after merging this first version though.

File: .github/workflows/tests.yml (hunk 2 above)
Comment (julienfalque) [reply .github/workflows/tests.yml:46]: True. What about `--no-interaction`? Can we remove it?

**Comment by julienfalque** [comment]
Shouldn't the jobs appear in the checks of this PR?

**Comment by GrahamCampbell** [comment]
Related: https://github.com/PHPUnitGoodPractices/Traits/pull/48.

**Comment by GrahamCampbell** [comment]
> Shouldn't the jobs appear in the checks of this PR?

No, not until this is merged.

File: .appveyor.yml [hunk 7]
```diff
@@ -1,45 +0,0 @@
-build: false
```
Comment (keradus) [inline .appveyor.yml:1]: purpose of AppVeyor was to test PHP CS Fixer under Windows too, instead of only unix. please keep AppVeyor

**Comment by GrahamCampbell** [comment]
Looks like the accessible object package is causing us issues. We need to mark it as compatible with PHP 8 and remove the HHVM conflict.

**Comment by GrahamCampbell** [comment]
See https://github.com/PHP-CS-Fixer/AccessibleObject/pull/3.

**Comment by GrahamCampbell** [comment]
Looks like these tests have actually highlighted issues with our minimum symfony console versions.

**Comment by GrahamCampbell** [comment]
This is passing, and could be merged right away. A follow up PR can remove the rest of the travis jobs.

**Comment by GrahamCampbell** [comment]
https://github.com/GrahamCampbell/PHP-CS-Fixer/actions/runs/376358990

**Comment by julienfalque** [comment]
@GrahamCampbell The PR is still a draft, shall we mark it as ready for review?

**Comment by keradus** [comment]
@GrahamCampbell , can you talk to creator of #5268 and align how to achieve best outcome together ;)
We have 2 competitive PRs, let's avoid duplicating the effort in work on the job separately.

File: tests/AutoReview/TravisTest.php [hunk 8]
```diff
@@ -1,184 +0,0 @@
...
-/**
- * @author Dariusz Rumiński <dariusz.ruminski@gmail.com>
- *
- * @internal
- *
- * @coversNothing
- * @group auto-review
- * @group covers-nothing
- */
-final class TravisTest extends TestCase
```
Comment (keradus) [inline tests/AutoReview/TravisTest.php:30]: :-1: for dropping
i would rather expecting to have those tests migrated to read github actions config than fully dropped.

File: .github/workflows/tests.yml [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        php: ["5.6", "7.1", "7.4"]
+        stability: ["Stable"]
+
+    name: PHP ${{ matrix.php }} - ${{ matrix.stability }} - Windows
+
+    steps:
+      - name: Set git to use LF
+        run: |
+          git config --global core.autocrlf false
+          git config --global core.eol lf
```
Comment (sanmai) [inline .github/workflows/tests.yml:93]: Why is this required?

I haven't seen anything bad happen without these lines.

**Review by sanmai — COMMENTED** [review]
This is a fine PR, but since it has a lot if `if:`s,  and separate jobs for different OSes (where it should have just used matrix), it'll be a headache to support.

File: .github/workflows/tests.yml (hunk 9 above)
Comment (GrahamCampbell) [reply .github/workflows/tests.yml:93]: Tests fail without I think.

File: .github/workflows/tests.yml [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    - cron: '0 0 * * *'
+
+jobs:
+  linux:
+
+    runs-on: ubuntu-20.04
+    strategy:
+      matrix:
+        php: ["5.6", "7.0", "7.1", "7.2", "7.3", "7.4"]
+        stability: ["Stable"]
```
Comment (kubawerlos) [inline .github/workflows/tests.yml:16]: What is the profit of using `stability` and the then triplicated "Install dependencies" with if?

Why can't we have key like `composer_flags: '--prefer-lowest'` and use it in the command?

File: .github/workflows/tests.yml [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Setup PHP
+        uses: shivammathur/setup-php@v2
+        with:
+          php-version: ${{ matrix.php }}
+          tools: composer:snapshot
```
Comment (kubawerlos) [inline .github/workflows/tests.yml:33]: What is `composer:snapshot` for? Why can't we use the latest stable Composer which cames by default?

File: .github/workflows/tests.yml (hunk 3 above)
Comment (kubawerlos) [reply .github/workflows/tests.yml:68]: Add the env in the matrix for PHP 8.0 then, so 2 steps for tests won't be needed.

File: .github/workflows/tests.yml [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+      - name: Execute tests
+        run: vendor/bin/phpunit
+        if: "matrix.php != '8.0'"
+
+      - name: Execute tests with weak deprecation warnings
+        run: SYMFONY_DEPRECATIONS_HELPER=weak vendor/bin/phpunit
+        if: "matrix.php == '8.0'"
+
+      - name: Execute php-cs-fixer standard mode
+        run: PHP_CS_FIXER_FUTURE_MODE=1 php php-cs-fixer --diff --dry-run -v fix
```
Comment (kubawerlos) [inline .github/workflows/tests.yml:72]: Same here, add needed env in matrix instead of duplicating steps just for an env.

File: .github/workflows/tests.yml (hunk 9 above)
Comment (kubawerlos) [reply .github/workflows/tests.yml:93]: Haven't seen this in other repos - can we try without this? Would be easier not to worry that this might affect some tests.

File: .github/workflows/tests.yml [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+        uses: actions/checkout@v2
+
+      - name: Setup PHP
+        uses: shivammathur/setup-php@v2
+        with:
+          php-version: ${{ matrix.php }}
+          tools: composer:snapshot
+          coverage: none
+
+      - name: Setup problem matchers
```
Comment (kubawerlos) [inline .github/workflows/tests.yml:36]: What is this? Can you elaborate or link to some PR when it is run or better - failing a build for good reason?

File: .github/workflows/tests.yml [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+
+      - name: Execute php-cs-fixer standard mode
+        run: PHP_CS_FIXER_FUTURE_MODE=1 php php-cs-fixer --diff --dry-run -v fix
+        if: "matrix.php != '8.0'"
+
+      - name: Execute php-cs-fixer ignoring env
+        run: PHP_CS_FIXER_FUTURE_MODE=1 PHP_CS_FIXER_IGNORE_ENV=1 php php-cs-fixer --diff --dry-run -v fix
+        if: "matrix.php == '8.0'"
+
+  windows:
```
Comment (kubawerlos) [inline .github/workflows/tests.yml:79]: Why do we need this? Why OS cannot be part of matrix?

**Review by kubawerlos — CHANGES_REQUESTED** [review]

File: .github/workflows/tests.yml [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+on:
+  push:
+  pull_request:
+  schedule:
+    - cron: '0 0 * * *'
+
+jobs:
+  linux:
+
+    runs-on: ubuntu-20.04
```
[the remark this answers is not in the document]
Comment (GrahamCampbell) [reply .github/workflows/tests.yml:12]: Because latest changes. I would consider it best practice to pin the version we want, to avoid breakages in the future.

File: .github/workflows/tests.yml (hunk 10 above)
Comment (GrahamCampbell) [reply .github/workflows/tests.yml:16]: Because this way we can easily render what is going on in the GitHub Actions UI for job descriptions. Much more important that those are readable, than the config file is.

File: .github/workflows/tests.yml (hunk 11 above)
Comment (GrahamCampbell) [reply .github/workflows/tests.yml:33]: Whoops, that can be changed to just `composer` now.

File: .github/workflows/tests.yml (hunk 13 above)
Comment (GrahamCampbell) [reply .github/workflows/tests.yml:36]: Shows test failures inline on diffs on github's ui.

File: .github/workflows/tests.yml (hunk 14 above)
Comment (GrahamCampbell) [reply .github/workflows/tests.yml:79]: Each OS is different, and in the end, trying to put it all into one single thing ends up an absolute mess. Sometimes duplication is least worst. Certainly makes the config file more readable.

File: .github/workflows/tests.yml (hunk 12 above)
Comment (GrahamCampbell) [reply .github/workflows/tests.yml:70]: I feel this is way makes the config file the least messy, and understandable.

**Comment by GrahamCampbell** [comment]
https://github.com/GrahamCampbell/PHP-CS-Fixer/actions/runs/380065972

File: tests/AutoReview/TravisTest.php (hunk 8 above)
Comment (GrahamCampbell) [reply tests/AutoReview/TravisTest.php:30]: Do you consider this a blocker for merging. I feel like the GitHub UI already makes it pretty clear if stuff is missing, with the human readable job names I have gone for.

**Comment by GrahamCampbell** [comment]
https://github.com/GrahamCampbell/PHP-CS-Fixer/actions/runs/380086877

File: .github/workflows/tests.yml (hunk 13 above)
Comment (sanmai) [reply .github/workflows/tests.yml:34]: Is it? I don't think I can reproduce: https://github.com/sanmai/phpunit-primer/pull/6/files

Comment (GrahamCampbell) [reply .github/workflows/tests.yml:34]: Yes, that stuff works properly. It's exactly as documented in https://github.com/shivammathur/setup-php#phpunit.

Comment (sanmai) [reply .github/workflows/tests.yml:34]: Well, the thing is I made [a test to fail intentionally](https://github.com/sanmai/phpunit-primer/pull/6/files#diff-13e3993e50d053aa5752f11a1e9bbf312e598ffa0054b8740260e29b7932d821R15-R18), added [these matchers](https://github.com/sanmai/phpunit-primer/blob/5d8c8bcbc5794b536c2bb9860c08866c21e295dc/.github/workflows/ci.yaml#L68), and there's nothing. How can we verify that this thing actually works?

File: .github/workflows/tests.yml [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Setup PHP
+        uses: shivammathur/setup-php@v2
+        with:
+          php-version: ${{ matrix.php }}
+          tools: composer
```
Comment (sanmai) [inline .github/workflows/tests.yml:31]: This is as redundant as before.

File: .github/workflows/tests.yml (hunk 9 above)
Comment (sanmai) [reply .github/workflows/tests.yml:91]: Are you sure? I have a workflow which [does not fail without this](https://github.com/sanmai/PHP-CS-Fixer/actions/runs/378021524) on Windows.

File: tests/AutoReview/TravisTest.php (hunk 8 above)
Comment (keradus) [reply tests/AutoReview/TravisTest.php:30]: yes, I do

but pls, align with `@sanmai` to prepare single PR instead of putting more effort in one or another PR

File: .github/workflows/tests.yml (hunk 13 above)
Comment (keradus) [reply .github/workflows/tests.yml:34]: @GrahamCampbell why closing the thread? looks like the question is still open

**Comment by GrahamCampbell** [comment]
If someone else wants to finish this, they should take over from here.

File: .github/workflows/tests.yml (hunk 14 above)
Comment (keradus) [reply .github/workflows/tests.yml:77]: not sure, here it was added without that "absolute mess"
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5268/files#diff-944291df2c9c06359d37cc8833d182d705c9e8c3108e7cfe132d61a06e9133ddR33

**Comment by keradus** [comment]
Thanks for your effort on this PR, @GrahamCampbell .
It's a pity that we didn't manage to have single PR for GH actions.
Closing in favor of #5268, but let's treat #5268 as a starting point where we can all improve further, same time having CI for repo running
